cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -597,19 +602,149 @@ public void getStatuses(HttpRequest request, HttpResponder responder,
       } catch (NotFoundException e) {
         statuses.add(new BatchProgramStatus(
-          program, HttpResponseStatus.NOT_FOUND.getCode(), null, e.getMessage()));
+          program, HttpResponseStatus.NOT_FOUND.getCode(), e.getMessage(), null));
       }
     }
     responder.sendJson(HttpResponseStatus.OK, statuses);
   }
 
   /**
+   * Stops all programs that are passed into the data. The data is an array of JSON objects
+   * where each object must contain the following three elements: appId, programType, and programId
+   * (flow name, service name, etc.).
+   * <p>
+   * Example input:
+   * <pre><code>
+   * [{"appId": "App1", "programType": "Service", "programId": "Service1"},
+   * {"appId": "App1", "programType": "Mapreduce", "programId": "MapReduce2"},
+   * {"appId": "App2", "programType": "Flow", "programId": "Flow1"}]
+   * </code></pre>
+   * </p><p>
+   * The response will be an array of JsonObjects each of which will contain the three input parameters
+   * as well as a "statusCode" field which maps to the status code for the data in that JsonObjects.
+   * </p><p>
+   * If an error occurs in the input (for the example above, App2 does not exist), then all JsonObjects for which the
+   * parameters have a valid status will have the status field but all JsonObjects for which the parameters do not have
+   * a valid status will have an error message and statusCode.
+   * </p><p>
+   * For example, if there is no App2 in the data above, then the response would be 200 OK with following possible data:
+   * </p>
+   * <pre><code>
+   * [{"appId": "App1", "programType": "Service", "programId": "Service1", "statusCode": 200},
+   * {"appId": "App1", "programType": "Mapreduce", "programId": "Mapreduce2", "statusCode": 200},
+   * {"appId":"App2", "programType":"Flow", "programId":"Flow1", "statusCode":404, "error": "App: App2 not found"}]
+   * </code></pre>
+   */
+  @POST
+  @Path("/stop")
+  public void stopPrograms(HttpRequest request, HttpResponder responder,
+                           @PathParam("namespace-id") String namespaceId) throws BadRequestException, IOException {
+
+    List<BatchProgram> programs = validateAndGetBatchInput(request, BATCH_PROGRAMS_TYPE);
+
+    List<ListenableFuture<BatchProgramResult>> issuedStops = new ArrayList<>(programs.size());
+    for (final BatchProgram program : programs) {
+      Id.Program programId = Id.Program.from(namespaceId, program.getAppId(),
+                                             program.getProgramType(), program.getProgramId());
+      try {
+        ListenableFuture<BatchProgramResult> issuedStop = Futures.transform(issueStop(programId, null),
+          new Function<ProgramController, BatchProgramResult>() {
+            @Override
+            public BatchProgramResult apply(ProgramController input) {
+              return new BatchProgramResult(
+                program, AppFabricServiceStatus.OK.getCode().getCode(), AppFabricServiceStatus.OK.getMessage());
```

Comment:
we should pass null as the last parameter to the constructor, right?
Since the last parameter is the error string.